usr/sbin/p11sak/p11sak.c

```diff
@@ -1185,101 +1237,120 @@ static CK_RV print_vendor(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hkey, int
+                    } else if (strcmp((confignode_to_bareval(type)->value), "CK_BYTE") == 0) {
+                        // get length 
+                        CK_ATTRIBUTE temp = {hex, NULL, 0};
+                        rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+                        if (rc == CKR_ATTRIBUTE_SENSITIVE && long_print) {
+                            printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                        } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID && rc != CKR_OK) {
+                            goto done;
+                        } else if (rc == CKR_OK && temp.ulValueLen != CK_UNAVAILABLE_INFORMATION) {
+                            // build template
+                            unsigned long a_byte_array = temp.ulValueLen;
+                            temp.pValue = malloc(a_byte_array);
+                            if (!temp.pValue) {
+                                fprintf(stderr, "Error: cannot malloc storage for attribute value.\n");
+                                rc = CKR_HOST_MEMORY;
+                                goto done;
+                            }
+
+                            // get attribute value
+                            rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+                            if (rc == CKR_OK) {
+                                if (temp.ulValueLen != a_byte_array) {
+                                    fprintf(stderr, " Error in retrieving Attribute %s: %lu\n",
+                                            confignode_to_bareval(name)->value, temp.ulValueLen);
+                                } else {
+                                    if (long_print) {
+                                        dump_attr(&temp);
+                                    } else {
+                                        printf(" 1 ");
+                                        free(temp.pValue);
+                                        rc = CKR_OK;
+                                        goto done;
+                                    }
+                                }    
+                            } 
+                            free(temp.pValue);
+                        } 
+                    } else {
+                        fprintf(stderr, "Error: Attribute type [%s] invalid.\n", confignode_to_bareval(type)->value);
+                    }
+                    attrs[index] = hex; // add attr as printed
                 } else {
-                    fprintf(stderr, "Error: Attribute type [%s] invalid.\n", confignode_to_bareval(type)->value);
+                    duplicate_attrs_in_conf = 1;
                 }
             } 
+            index++;
         }
     }
     if (!long_print)
         printf(" %s ", def_attr ? "0" : "-");
     
-    return CKR_OK;
+    rc = CKR_OK;
+
+done:
+    if (duplicate_attrs_in_conf)
+        fprintf(stderr, "Warning: There are duplicate vendor attr definitions in config file, please check.\n");
```

Comment:
same here. Please don't call this 'vendor' attributes. Maybe also rename the functions you added to something else than 'vendor'.

Reply:
Done.